### PR TITLE
L/Q again, step two

### PR DESCRIPTION
**Pull request by fparain**
Please review this second set of fixes for the L/Q transition.
A few changes in C1, most changes in CI.
CI always delegates to the runtime to know if a null-free array is flattened or not (to avoid duplication of the flattening decision code).

With those changes, all tests in runtime/valhalla/inlinetypes now pass with -Xcomp -XX:TieredStopAtLevel=1 (at least on Linux x64).
Some tests in compiler/valhalla/inlinetypes still fail but they will require more time to be investigated (some tests force use of C2 or expect code to be C2 compiled, but C2 is not fixed yet).

Thank you,

Fred

---------
### Progress
- [x] Change must not contain extraneous whitespace

### Reviewers
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/valhalla pull/414/head:pull/414` \
`$ git checkout pull/414`

Update a local copy of the PR: \
`$ git checkout pull/414` \
`$ git pull https://git.openjdk.java.net/valhalla pull/414/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 414`

View PR using the GUI difftool: \
`$ git pr show -t 414`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/valhalla/pull/414.diff">https://git.openjdk.java.net/valhalla/pull/414.diff</a>

</details>

### CODE REVIEW

File: test/hotspot/jtreg/runtime/valhalla/inlinetypes/MultiANewArrayTest/MultiANewArrayTest.java [hunk 1]
```diff
@@ -49,8 +49,7 @@ public static void main(String[] args) {
         } catch(Error e) {
             ex = e;
         }
-        Asserts.assertNotNull(ex, "An ICCE should have been thrown");
-        Asserts.assertEquals(ex.getClass(), IncompatibleClassChangeError.class, "Error is not an ICCE");
+        Asserts.assertNull(ex, "No ICCE should have been thrown");
```
Comment (forax) [inline test/hotspot/jtreg/runtime/valhalla/inlinetypes/MultiANewArrayTest/MultiANewArrayTest.java:52]: The negation of "An ICCE should have been thrown" is "No error should have been thrown",
talking about ICCE it that context can be misleading.

Comment (fparain) [reply test/hotspot/jtreg/runtime/valhalla/inlinetypes/MultiANewArrayTest/MultiANewArrayTest.java]: Error message updated in the last commit.

Thanks,

Fred

File: src/hotspot/cpu/x86/c1_CodeStubs_x86.cpp [hunk 2]
```diff
@@ -310,12 +310,12 @@ void NewTypeArrayStub::emit_code(LIR_Assembler* ce) {
 // Implementation of NewObjectArrayStub
 
 NewObjectArrayStub::NewObjectArrayStub(LIR_Opr klass_reg, LIR_Opr length, LIR_Opr result,
-                                       CodeEmitInfo* info, bool is_inline_type) {
+                                       CodeEmitInfo* info, bool is_null_free) {
```
Comment (nick-arm) [inline src/hotspot/cpu/x86/c1_CodeStubs_x86.cpp:313]: Would you mind making the same change in `c1_CodeStubs_aarch64.cpp` to keep them in sync?

Comment (fparain) [reply src/hotspot/cpu/x86/c1_CodeStubs_x86.cpp:313]: Sure, sorry for the breakage of the aarch64 platform.

Fred

File: src/hotspot/share/ci/ciEnv.cpp [hunk 3]
```diff
@@ -475,7 +475,8 @@ ciKlass* ciEnv::get_klass_by_name_impl(ciKlass* accessing_klass,
                              require_local);
     if (elem_klass != NULL && elem_klass->is_loaded()) {
       // Now make an array for it
-      return ciArrayKlass::make(elem_klass);
+      bool null_free_array = sym->is_Q_array_signature() && sym->char_at(1) == JVM_SIGNATURE_INLINE_TYPE;
```
Comment (TobiHartmann) [inline src/hotspot/share/ci/ciEnv.cpp:478]: Why do we need the `sym->char_at(1) == JVM_SIGNATURE_INLINE_TYPE` check? Isn't that part of `is_Q_array_signature()`?

File: src/hotspot/share/ci/ciArrayKlass.cpp [hunk 4]
```diff
@@ -104,10 +105,36 @@ bool ciArrayKlass::is_leaf_type() {
...
+          }
+        }
+      )
+      if (is_array_flattened) {
+        return ciFlatArrayKlass::make(klass);
+      } else {
+        return ciObjArrayKlass::make(klass, true);
+      }
+    } else {
+      return ciEnv::unloaded_ciobjarrayklass();
```
Comment (TobiHartmann) [inline src/hotspot/share/ci/ciArrayKlass.cpp:134]: Can't we delegate handling the unloaded case to `ciObjArrayKlass::make(klass)` below? I.e. check for `if (null_free && klass->is_loaded())` above.

File: src/hotspot/share/ci/ciArrayKlass.cpp [hunk 5]
```diff
@@ -104,10 +105,36 @@ bool ciArrayKlass::is_leaf_type() {
...
+    if (klass->is_loaded()) {
+      bool is_array_flattened = false;
+      GUARDED_VM_ENTRY(
+        EXCEPTION_CONTEXT;
+        Klass* ak = InlineKlass::cast(klass->get_Klass())->null_free_inline_array_klass(THREAD);
+        if (HAS_PENDING_EXCEPTION) {
+          CLEAR_PENDING_EXCEPTION;
+        } else {
+          if (ak != NULL && ak->is_flatArray_klass()) {
+            is_array_flattened = true;
```
Comment (TobiHartmann) [inline src/hotspot/share/ci/ciArrayKlass.cpp:124]: Why is that boolean needed? Can't we simply return `ciFlatArrayKlass::make(klass)` here?

**Review by TobiHartmann — CHANGES_REQUESTED** [review]
Looks good to me. Added some minor comments.

File: src/hotspot/share/ci/ciEnv.cpp (hunk 3 above)
Comment (fparain) [reply src/hotspot/share/ci/ciEnv.cpp:478]: The problem with is_Q_array_signature() is that it returns true for arrays of any dimension with the last dimension element klass being a null-free primitive class. It returns true for [QFoo; but also for [[[QFoo;. But for the [[[QFoo; array, the first two dimensions are regular nullable object arrays, only the last array is a null-free array (possibly flattened).

File: src/hotspot/share/ci/ciArrayKlass.cpp (hunk 4 above)
Comment (fparain) [reply src/hotspot/share/ci/ciArrayKlass.cpp]: I can change that. This was the way it was implemented initially, then I changed it to have the whole logic in a single method, but I don't have a strong opinion about this one.
Updated in the latest commit.

File: src/hotspot/share/ci/ciArrayKlass.cpp (hunk 5 above)
Comment (fparain) [reply src/hotspot/share/ci/ciArrayKlass.cpp]: Boolean removed in the latest commit.

File: src/hotspot/share/ci/ciArrayKlass.cpp [hunk 6]
```diff
@@ -104,10 +105,27 @@ bool ciArrayKlass::is_leaf_type() {
...
+
+ciArrayKlass* ciArrayKlass::make(ciKlass* klass, bool null_free) {
+  if (null_free && klass->is_loaded()) {
+    GUARDED_VM_ENTRY(
+      EXCEPTION_CONTEXT;
+      Klass* ak = InlineKlass::cast(klass->get_Klass())->null_free_inline_array_klass(THREAD);
+      if (HAS_PENDING_EXCEPTION) {
+        CLEAR_PENDING_EXCEPTION;
+      } else {
+        if (ak != NULL && ak->is_flatArray_klass()) {
```
Comment (TobiHartmann) [inline src/hotspot/share/ci/ciArrayKlass.cpp:121]: `else` and `if` can be merged

File: src/hotspot/share/ci/ciEnv.cpp (hunk 3 above)
Comment (TobiHartmann) [reply src/hotspot/share/ci/ciEnv.cpp:478]: Ah, right. Thanks for the explanation!

**Review by TobiHartmann — APPROVED** [review]
Thanks for updating, looks good to me (added one minor suggestion).

File: src/hotspot/share/ci/ciArrayKlass.cpp (hunk 6 above)
Comment (fparain) [reply src/hotspot/share/ci/ciArrayKlass.cpp]: Code updated.

**Review by TobiHartmann — APPROVED** [review]

**Pull request closed without merging**